dvc/remote/index.py

```diff
@@ -0,0 +1,84 @@
+import logging
+import os
+import pickle
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteIndex(object):
+    """Class for locally indexing remote checksums.
+
+    Args:
+        repo: repo for this remote index.
+        name: name for this index. If name is provided, this index will be
+            loaded from and saved to ``.dvc/tmp/index/{name}.idx``.
+            If name is not provided (i.e. for local remotes), this index will
+            be kept in memory but not saved to disk.
+    """
+
+    INDEX_SUFFIX = ".idx"
+
+    def __init__(self, repo, name=None):
+        if name:
+            self.path = os.path.join(
+                repo.index_dir, "{}{}".format(name, self.INDEX_SUFFIX)
+            )
+        else:
+            self.path = None
+        self._checksums = set()
+        self.load()
+
+    def __iter__(self):
+        return iter(self._checksums)
+
+    @property
+    def checksums(self):
+        return self._checksums
+
+    def load(self):
+        """(Re)load this index from disk."""
+        if self.path and os.path.isfile(self.path):
+            try:
+                with open(self.path, "rb") as fobj:
+                    self._checksums = pickle.load(fobj)
+            except PermissionError:
```

Comment:
why do we catch only PermissionErrors? not IO in general?

Reply:
there was an earlier review comment to catch more specific exceptions, I'll fix this to also catch general IOError as well

Reply:
I wonder if we should at least show some message (let's not use WARNING though, please - they look ugly) in some cases? To avoid situation when some logic is broken or not enough permissions (can it happen at all?) and we silently degrade performance significantly.

Reply:
I think the only time IO errors would actually happen here is in the event we can't read/write to the `.dvc/tmp` directory at all, whether it's due to permissions, lack of disk space, or some other reason (which I think would cause errors elsewhere in dvc).

Reply:
Updated this so that on an IO error we explicitly state that the remote will be re-indexed, and the message includes the original exception/error message.

It's still logged as an error though, should this be an info message since it's not fatal?